### PR TITLE
v11 C++ Changes

### PR DESCRIPTION
**Pull request by carrus85**
## Description

Includes:
* v11 C++ format changes, which introduces a per-file "metadata" section for storing additional information about a file.
* v10a unconverter changes, which emits metadata lines indicating which columns are "non-empty" on a block by block basis optionally.
* the `getLineage` function, built on top of the v11 C++ changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: cplusplus/zdw/UnconvertFromZDW.h [hunk 1]
```diff
@@ -19,13 +19,13 @@
...
 
 #include <map>
+#include <set>
 #include <string>
 #include <vector>
 #include <sys/stat.h>
 
 #include <boost/scoped_ptr.hpp>
 
-
```
Comment (scotthunt) [inline cplusplus/zdw/UnconvertFromZDW.h]: nit: To match every other file, perhaps leave the double blank line here?

File: cplusplus/zdw/UnconvertFromZDW.h [hunk 2]
```diff
@@ -83,8 +84,14 @@ struct VisitorPart
 	internal::indexBytes m_PrevID;
 };
 
-} // namespace internal
+struct MetadataOptions {
```
Comment (scotthunt) [inline cplusplus/zdw/UnconvertFromZDW.h]: nit: all other structs have their open brace on the next line

File: cplusplus/zdw/UnconvertFromZDW.h [hunk 3]
```diff
@@ -115,18 +122,26 @@ class UnconvertFromZDW_Base
 
 	void printError(const std::string &exeName, const std::string &inFileName);
 
+	void outputNonEmptyColumnHeader(bool bFlag=true) { this->bOutputNonEmptyColumnHeader = bFlag; }
```
Comment (scotthunt) [inline cplusplus/zdw/UnconvertFromZDW.h]: nit: all other default args have spaces around the '='

File: cplusplus/zdw/UnconvertFromZDW.h [hunk 4]
```diff
@@ -115,18 +122,26 @@ class UnconvertFromZDW_Base
...
 	ERR_CODE readHeader();
 	bool setNamesOfColumnsToOutput(const std::string& csv_str, COLUMN_INCLUSION_RULE inclusionRule);
 	bool setNamesOfColumnsToOutput(const std::vector<std::string> &csv_vector, COLUMN_INCLUSION_RULE inclusionRule);
 	void showBasicStatisticsOnly(bool bVal = true) { this->bShowBasicStatisticsOnly = bVal; }
 
 	ERR_CODE GetSchema(std::ostream& stream);
 
+	void setMetadataOptions(const internal::MetadataOptions& options) {
```
Comment (scotthunt) [inline cplusplus/zdw/UnconvertFromZDW.h]: nit: all other [multi-line] functions have their open braces on the next line

File: cplusplus/ConvertToZDW.cpp [hunk 5]
```diff
@@ -49,7 +52,8 @@ using std::vector;
 //version 9a -- added support for passing in arguments to the file compression process
 //version 9b -- Allowing > 2040 columns to be set in a block (via fixing an integer overflow)
 //version 10 -- add support for mediumtext and longtext columns
-
+//version 10a -- version 11 support built in (use of v11 is disabled by default)
+//version 11 -- add metadata block to file header
 
 namespace {
```
Comment (scotthunt) [inline cplusplus/ConvertToZDW.cpp:59]: nit: lost double newline before namespace

File: cplusplus/ConvertToZDW.cpp [hunk 6]
```diff
@@ -214,6 +219,45 @@ ConvertToZDW::ERR_CODE ConvertToZDW::validate(
 	return err == 0 ? OK : FILES_DIFFER;
 }
 
+//Returns: whether inputted metadata is valid
+bool ConvertToZDW::validateMetadata(const map<string, string>& metadata) const
+{
+	for (map<string, string>::const_iterator it=metadata.begin(); it!=metadata.end(); ++it) {
```
Comment (scotthunt) [inline cplusplus/ConvertToZDW.cpp]: nit: missing spaces around operators (both '=' and '!=')

File: cplusplus/ConvertToZDW.cpp [hunk 7]
```diff
@@ -214,6 +219,45 @@ ConvertToZDW::ERR_CODE ConvertToZDW::validate(
...
+int ConvertToZDW::loadMetadataFile(const char* filepath,
+	map<string, string>& metadata) //(in/out)
+{
+	std::ifstream stream(filepath);
+
+	if (!stream)
+		return -1;
+
+	string line;
+	for (int linenum=1; getline(stream, line); ++linenum)
```
Comment (scotthunt) [inline cplusplus/ConvertToZDW.cpp]: bit: missing spaces around operator '='

File: cplusplus/ConvertToZDW.cpp [hunk 8]
```diff
@@ -214,6 +219,45 @@ ConvertToZDW::ERR_CODE ConvertToZDW::validate(
...
+	{
+		if (line.empty())
+			continue;
+
+		const size_t position = line.find('=');
+		if (position == string::npos)
+			return linenum;
+
+		metadata[line.substr(0, position)] = line.substr(position+1);
```
Comment (scotthunt) [inline cplusplus/ConvertToZDW.cpp]: bit: missing spaces around '+' operator

File: cplusplus/ConvertToZDW.cpp [hunk 9]
```diff
@@ -621,6 +676,27 @@ ConvertToZDW::ERR_CODE ConvertToZDW::processFile(
 	//Write version #.
 	fwrite(&m_Version, 1, 2, out);
 
+	//Write metadata block.
+	if (CONVERT_ZDW_CURRENT_VERSION == 11) //TODO temp feature flag hack for migration from version 10 to version 11
+	{
+		ULONG metadata_length = 0;
+		map<string, string>::const_iterator it;
+		for (it=metadata.begin(); it!=metadata.end(); ++it) {
```
Comment (scotthunt) [inline cplusplus/ConvertToZDW.cpp]: nit: missing spaces around '=' and '!='

File: cplusplus/ConvertToZDW.cpp [hunk 10]
```diff
@@ -621,6 +676,27 @@ ConvertToZDW::ERR_CODE ConvertToZDW::processFile(
...
+			metadata_length += it->first.length();
+			metadata_length += it->second.length();
+			metadata_length += 2; //two null terminators
+		}
+		fwrite(&metadata_length, 1, 4, out);
+
+		for (it=metadata.begin(); it!=metadata.end(); ++it) {
```
Comment (scotthunt) [inline cplusplus/ConvertToZDW.cpp]: nit: missing spaces around '=' and '!='

File: cplusplus/ConvertToZDW.cpp [hunk 11]
```diff
@@ -918,6 +995,16 @@ ConvertToZDW::ERR_CODE ConvertToZDW::convertFile(
...
+	map<string, string> inMetadata = metadata;
+	if (inMetadata.empty() && CONVERT_ZDW_CURRENT_VERSION >= 11) {
+		//Default to .metadata file contents, if present.
+		sprintf(command, "%s.metadata", filestub);
+		const int res = loadMetadataFile(command, inMetadata);
+		if (res > 0) //ignore -1: it's okay for file to be not present
+			return BAD_METADATA_FILE;
+	}
+
+
```
Comment (scotthunt) [inline cplusplus/ConvertToZDW.cpp]: nit: extra blank line?

File: cplusplus/convertDWfile.cpp [hunk 12]
```diff
@@ -145,10 +156,32 @@ int main(int argc, char* argv[])
 								return badParam(program, argv[i]);
 							break;
 						}
+						if (!strncmp(flag, "metadata:", 9)) {
+							const char *key = flag+9;
```
Comment (scotthunt) [inline cplusplus/convertDWfile.cpp]: nit: missing spaces around '+'

File: cplusplus/convertDWfile.cpp [hunk 13]
```diff
@@ -145,10 +156,32 @@ int main(int argc, char* argv[])
...
+							if (!value)
+								return badParam(program, argv[i]);
+							metadata[string(key, value - key)] = string(value + 1);
+							break;
+						}
+						if (!strncmp(flag, "metadata-file=", 14)) {
+							const char *filepath = flag + 14;
+							const int line_res = ConvertToZDW::loadMetadataFile(filepath, metadata);
+							if (line_res) {
+								fprintf(stderr, "%s: Metadata file load error '%s' (line %d) \n\n", program, filepath, line_res);
```
Comment (scotthunt) [inline cplusplus/convertDWfile.cpp]: nit: extra space at end of line? (before newline)

File: cplusplus/unconvertDWfile.cpp [hunk 14]
```diff
@@ -63,6 +63,17 @@ void usage(char* executable)
 	       "\t-t test integrity of zdw file only\n"
 	       "\t-v verbose -- show count of rows during conversion\n"
 	       "\t-w give outputted files no extension (default = .sql)\n"
+		   "\n"
```
Comment (scotthunt) [inline cplusplus/unconvertDWfile.cpp]: nit: whitespace mismatch against previous lines?

File: cplusplus/UnconvertFromZDW.cpp [hunk 15]
```diff
@@ -129,7 +132,7 @@ void InitDirAndBasenameFromFileName(string const& inFileNameStr, char * &sourceD
 namespace adobe {
 namespace zdw {
 
-const int UnconvertFromZDW_Base::UNCONVERT_ZDW_VERSION = 10;
+const int UnconvertFromZDW_Base::UNCONVERT_ZDW_VERSION = 11;
 const char UnconvertFromZDW_Base::UNCONVERT_ZDW_VERSION_TAIL[3] = "";
 
 const size_t UnconvertFromZDW_Base::DEFAULT_LINE_LENGTH = 16*1024; //16K default
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: looks like I missed one in my previous whitespace cleanup!

File: cplusplus/UnconvertFromZDW.cpp [hunk 16]
```diff
@@ -642,6 +642,71 @@ string UnconvertFromZDW_Base::getColumnDesc(const string& name, UCHAR type, size
...
+}
+
+ERR_CODE UnconvertFromZDW_Base::outputMetadataToStdOut() const
+{
+	return outputMetadata(stdout);
+}
+
+//****************************************************
+ERR_CODE UnconvertFromZDW_Base::outputMetadata(FILE *out)
+const
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: nit: const is on previous line on every other const function/method

File: cplusplus/UnconvertFromZDW.cpp [hunk 17]
```diff
@@ -642,6 +642,71 @@ string UnconvertFromZDW_Base::getColumnDesc(const string& name, UCHAR type, size
...
+{
+	const set<string>& keys = this->metadataOptions.keys;
+
+	//Filter output set
+	set<string> outKeys;
+	set<string>::const_iterator it;
+	for (it=keys.begin(); it!=keys.end(); ++it)
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: nit: missing whitespace around '=' and '!='

File: cplusplus/UnconvertFromZDW.cpp [hunk 18]
```diff
@@ -642,6 +642,71 @@ string UnconvertFromZDW_Base::getColumnDesc(const string& name, UCHAR type, size
...
+			this->metadata.find(*it) == this->metadata.end())
+		{
+			return METADATA_KEY_NOT_PRESENT;
+		}
+
+		outKeys.insert(*it);
+	}
+
+	//Output metadata keys
+	map<string,string>::const_iterator m_it;
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: nit: missing space after ','

File: cplusplus/UnconvertFromZDW.cpp [hunk 19]
```diff
@@ -926,12 +1015,36 @@ ERR_CODE UnconvertFromZDW_Base::readHeader()
...
+			this->metadata[string(key)] = string(value);
 
-	//2. Parse file attributes here (before version 3).
-	if (this->version <= 2)
+			//advance to next key-value pair
+			key = value + strlen(value) + 1; //skip past null terminator
+		}
+		free(metadata_block);
+	} else if (this->version <= 2)
 	{
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: nit: no other place has linebreaks around braces like this (i.e. "} ...\n{".  perhaps pull out the newline?  (though I admit, this is less bothersome to me... at least while out whitespace around control-flow braces is so schizophrenic)

File: cplusplus/UnconvertFromZDW.cpp [hunk 20]
```diff
@@ -1557,40 +1687,62 @@ ERR_CODE UnconvertFromZDWToFile<BufferedOutput_T>::unconvert(
...
-		char textbuf[1024];
-		sprintf(textbuf, "%s/%s%s", outputDir, outputBasename, ext ? ext : "");
-		if (this->bShowStatus)
-			this->statusOutput(INFO, "Writing %s\n", textbuf);
+		string outFileName = outputDir + "/" + outputBasename;
+		if (ext)
+			outFileName += ext;
+
+		if(this->bShowStatus)
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: nit: missing space before '('

File: cplusplus/UnconvertFromZDW.cpp [hunk 21]
```diff
@@ -1708,6 +1858,20 @@ ERR_CODE UnconvertFromZDWToMemory::getRow(char** buffer, size_t *size, const cha
 					return eRet;
 			}
 			break;
+			case ZDW_OUTPUT_BLOCK_HEADER:
+			{
+				assert(!this->rowsRead);
+
+				if(!bUseInternalBuffer) {
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: bit: missing space before '('

File: cplusplus/UnconvertFromZDW.cpp [hunk 22]
```diff
@@ -1878,12 +2047,51 @@ bool UnconvertFromZDWToMemory::OutputDescToFile(const string &outputDir)
...
 	const char* outputBasename = NULL;
 	InitDirAndBasenameFromFileName(this->inFileName, sourceDir, outputBasename);
-	ERR_CODE eRet = this->outputDescToFile(this->columnNames, outputDir.c_str(), outputBasename, ".sql");
+	ERR_CODE eRet = this->outputDescToFile(this->columnNames, outputDir, outputBasename, ".sql");
 	if (sourceDir)
 		free(sourceDir);
 	return eRet == OK;
 }
 
+vector<std::pair<uint64_t,string> > UnconvertFromZDWToMemory::getFileLineage()
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: nit: missing space after ','
also: this file already has "using std::pair", so perhaps remove the redundant "std::"?

File: cplusplus/UnconvertFromZDW.cpp [hunk 23]
```diff
@@ -1878,12 +2047,51 @@ bool UnconvertFromZDWToMemory::OutputDescToFile(const string &outputDir)
...
+{
+	const string LINEAGE_KEY = "lineage";
+
+	vector<std::pair<uint64_t,string> > out;
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: bit: missing space after ','

File: cplusplus/UnconvertFromZDW.cpp [hunk 24]
```diff
@@ -1878,12 +2047,51 @@ bool UnconvertFromZDWToMemory::OutputDescToFile(const string &outputDir)
...
+
+	if (this->eState == ZDW_BEGIN) {
+		ERR_CODE eRet = readHeader();
+		if (eRet != OK) {
+			out.push_back(std::make_pair<uint64_t,string>(0, string("bad ZDW header")));
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: bit: missing space after ','

File: cplusplus/UnconvertFromZDW.cpp [hunk 25]
```diff
@@ -1878,12 +2047,51 @@ bool UnconvertFromZDWToMemory::OutputDescToFile(const string &outputDir)
...
+			return out;
+		}
+	}
+
+	std::map<string, string>::const_iterator it = this->metadata.find(LINEAGE_KEY);
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: file already has "using std::map", perhaps remove redundant "std::"?

File: cplusplus/UnconvertFromZDW.cpp [hunk 26]
```diff
@@ -1878,12 +2047,51 @@ bool UnconvertFromZDWToMemory::OutputDescToFile(const string &outputDir)
...
+
+	const string &value = it->second;
+	if (value.empty())
+		return out;
+
+	size_t pos = 0, rowpos;
+	do {
+		rowpos = value.find(',', pos);
+		if (rowpos == string::npos) {
+			out.push_back(std::make_pair<uint64_t,string>(0, string("bad lineage data")));
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: nit: missing space after ','

File: cplusplus/UnconvertFromZDW.cpp [hunk 27]
```diff
@@ -1878,12 +2047,51 @@ bool UnconvertFromZDWToMemory::OutputDescToFile(const string &outputDir)
...
+			return out;
+		}
+
+		out.push_back(std::make_pair<uint64_t,string>(strtoull(value.c_str() + rowpos + 1, NULL, 10), value.substr(pos, rowpos - pos)));
```
Comment (scotthunt) [inline cplusplus/UnconvertFromZDW.cpp]: nit: missing space after ','

File: cplusplus/UnconvertFromZDW.cpp (hunk 15 above)
Comment (carrus85) [reply cplusplus/UnconvertFromZDW.cpp]: Missed what?  Care to clarify?

File: cplusplus/UnconvertFromZDW.cpp (hunk 19 above)
Comment (carrus85) [reply cplusplus/UnconvertFromZDW.cpp]: Adjusted to align with other uses of "else if" in the code.

File: cplusplus/UnconvertFromZDW.cpp (hunk 15 above)
Comment (carrus85) [reply cplusplus/UnconvertFromZDW.cpp]: Ah; whitespace around the multiply.  Fixed.

**Comment by scotthunt** [comment]
stuff I still see (if you care):
UnconvertFromZDW.cpp L#2058: redundant "std::"
memory.cpp L#22: spaces around '*'
ConvertToZDW.cpp L#256: spaces around '+'
ConvertToZDW.cpp L#975: spaces around '*'

**Review by mrimer — APPROVED** [review]

**Pull request merged**